networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/peers/PeerChainValidator.java

```diff
@@ -0,0 +1,197 @@
+    } else if (finalizedEpoch.compareTo(remoteFinalizedEpoch) > 0) {
+      // We're ahead of our peer, check that we agree with our peer's finalized epoch
+      verifyPeersFinalizedCheckpointIsCanonical(peerStatus, isChainValid);
+    } else {
+      // Our peer is ahead of us, check that they agree on our finalized epoch
+      verifyPeerAgreesWithOurFinalizedCheckpoint(finalizedCheckpoint, peerStatus, isChainValid);
+    }
+    return isChainValid;
+  }
+
+  private boolean verifyFinalizedCheckpointsAreTheSame(
+      Checkpoint finalizedCheckpoint, StatusData peerStatus) {
+    return Objects.equals(finalizedCheckpoint.getRoot(), peerStatus.getFinalizedRoot());
+  }
+
+  private void verifyPeersFinalizedCheckpointIsCanonical(
+      StatusData peerStatus, CompletableFuture<Boolean> isChainValid) {
+    final UnsignedLong remoteFinalizedEpoch = peerStatus.getFinalizedEpoch();
+    final UnsignedLong remoteFinalizedSlot = compute_start_slot_at_epoch(remoteFinalizedEpoch);
+    historicalChainData
+        // TODO - we need to get the block at or before this slot
+        .getBlockBySlot(remoteFinalizedSlot)
+        .thenApply(maybeBlock -> toBlock(remoteFinalizedSlot, maybeBlock))
+        .thenApply((block) -> validateBlockRootsMatch(block, peerStatus.getFinalizedRoot()))
+        .whenComplete(propagateFutureResult(isChainValid));
+  }
+
+  private void verifyPeerAgreesWithOurFinalizedCheckpoint(
+      Checkpoint finalizedCheckpoint,
+      StatusData peerStatus,
+      CompletableFuture<Boolean> isChainValid) {
+    final UnsignedLong finalizedEpochSlot =
+        compute_start_slot_at_epoch(finalizedCheckpoint.getEpoch());
+
+    historicalChainData
+        // TODO - we need to get the block at or before this slot
+        .getBlockBySlot(finalizedEpochSlot)
+        .thenApply(maybeBlock -> blockToSlot(finalizedEpochSlot, maybeBlock))
+        .thenCompose(blockSlot -> peer.requestBlockBySlot(peerStatus.getHeadRoot(), blockSlot))
+        .thenApply(block -> validateBlockRootsMatch(block, finalizedCheckpoint.getRoot()))
+        .whenComplete(propagateFutureResult(isChainValid));
+  }
+
+  private BeaconBlock toBlock(UnsignedLong lookupSlot, Optional<BeaconBlock> maybeBlock) {
+    return maybeBlock.orElseThrow(
+        () -> new IllegalStateException("Missing finalized block at slot " + lookupSlot));
+  }
+
+  private UnsignedLong blockToSlot(UnsignedLong lookupSlot, Optional<BeaconBlock> maybeBlock) {
+    if (maybeBlock.isEmpty()) {
+      throw new IllegalStateException("Missing historical block for slot " + lookupSlot);
+    }
+    return maybeBlock.get().getSlot();
+  }
+
+  private boolean validateBlockRootsMatch(final BeaconBlock block, final Bytes32 root) {
+    final Bytes32 blockRoot = block.signing_root("signature");
+    final boolean rootsMatch = Objects.equals(blockRoot, root);
+    if (rootsMatch) {
+      LOG.trace("Verified finalized blocks match for peer: {}", peer);
+    } else {
+      LOG.warn(
```

Comment:
This probably should just be trace level. It just means the peer is on a different chain to us and we should disconnect.

Reply:
Isn't the idea to disambiguate chains using the fork string?  In which case, a peer with the same fork string but an inconsistent finalized block means there was a soft fork or a consensus bug.  For now, I think it makes sense to be chatty in this case since we don't really expect this to be a regular occurrence and it could mean something bad is happening.

Reply:
Happy to see how often it pops up.  The fork doesn't identify anything about the genesis state so MainNet and Testnets will all be on the same fork but have completely different blocks.